update: conversion notes (#4838)
JetBrains/kotlin-web-site: docs/topics/native/native-objc-interop.md

>
{style="warning"}

To make your Kotlin code more Swift/Objective-C-friendly, you can hide a Kotlin declaration from Objective-C and Swift

> with `@HiddenFromObjC`. The annotation disables a function or property export to Objective-C.

Reviewer: To make your Kotlin code more Swift/Objective-C-friendly, use @HiddenFromObjC to hide a Kotlin declaration from Objective-C and Swift.

Reviewer: WDYT?

Alternatively, you can mark Kotlin declarations with the `internal` modifier to restrict their visibility in the

> compilation module. Choose `@HiddenFromObjC` if you only want to hide the Kotlin declaration from Objective-C and Swift

Reviewer: ```suggestion
compilation module. Use `@HiddenFromObjC` if you only want to hide the Kotlin declaration from Objective-C and Swift
```

Reviewer: Or maybe it is better to rephrase the whole sentence:

Use @HiddenFromObjC if you want to hide the Kotlin declaration from Objective-C and Swift while keeping it visible to other Kotlin modules.

that are handled at compile time. So, if Swift or Objective-C code calls a Kotlin method that throws an exception,
the Kotlin method should be marked with the `@Throws` annotation, specifying a list of "expected" exception classes.

When compiling to the Swift/Objective-C framework, non-`suspend` functions that have or inherit the `@Throws` annotation
are represented as `NSError*`-producing methods in Objective-C and as `throws` methods in Swift.
Representations for `suspend` functions always have an `NSError*`/`Error` parameter in the completion handler.

> When the Kotlin function called from Swift/Objective-C code throws an exception which is an instance of one of

Reviewer: ```suggestion
When a Kotlin function called from Swift/Objective-C code throws an exception which is an instance of one of
```

Reviewer: It's about any suspend Kotlin function, right?

> the `@Throws`-specified classes or their subclasses, it is propagated as `NSError`.

Reviewer: ```suggestion
the classes specified with `@Throws` or their subclasses, the exception is propagated as an `NSError`.
```

Reviewer: WDYT?

* [How to access Kotlin objects using `shared`](https://github.com/kotlin-hands-on/kotlin-swift-interopedia/blob/main/docs/classesandinterfaces/Objects.md)
* [How to access members of Kotlin companion objects from Swift](https://github.com/kotlin-hands-on/kotlin-swift-interopedia/blob/main/docs/classesandinterfaces/Companion%20objects.md).

### Primitive types

Kotlin primitive type boxes are mapped to special Swift/Objective-C classes. For example, the `kotlin.Int` box is represented

> as the `KotlinInt` class instance in Swift (or the `${prefix}Int` instance in Objective-C, where `prefix` is the framework names prefix).

Reviewer: ```suggestion
as the `KotlinInt` class instance in Swift (or the `${prefix}Int` instance in Objective-C, where `prefix` is the framework's name prefix).
```

Reviewer: Am I wrong here?

The `NSNumber` type is not automatically translated to Kotlin primitive types when used as a Swift/Objective-C parameter type
or return value. The reason is that the `NSNumber` type doesn't provide enough information about a wrapped primitive value
type, for example, `NSNumber` is statically not known to be `Byte`, `Boolean`, or `Double`. So Kotlin primitive values
should be [cast to and from `NSNumber` manually](#casting-between-mapped-types).

### Strings

When a Kotlin `String` is passed to Swift, it's first exported as an Objective-C object, and then the Swift compiler
copies it one more time for a Swift conversion. This results in additional runtime overhead.

To avoid that, access Kotlin strings in Swift directly as an Objective-C `NSString` instead.
[See the conversion example](#see-the-conversion-example).

#### NSMutableString

`NSMutableString` Objective-C class is not available from Kotlin.
All instances of `NSMutableString` are copied when passed to Kotlin.

### Collections

#### Kotlin -> Objective-C -> Swift

When a Kotlin collection is passed to Swift, it's first converted to an Objective-C equivalent, and then the Swift compiler

> copies the entire collection and converts it into a Swift-native collection as described in the [table above](#mappings).

Reviewer: ```suggestion
copies the entire collection and converts it into a Swift-native collection as described in the [mappings table](#mappings).
```

This last conversion leads to performance costs. To prevent this, when using Kotlin collections in Swift,
explicitly cast them to their Objective-C counterparts: `NSDictionary`, `NSArray`, or `NSSet`.

##### See the conversion example {initial-collapse-state="collapsed" collapsible="true"}

For example, the following Kotlin declaration:

```kotlin
data class Data(val map: Map<String, String>) {

>     fun getFromMap(key: String) = map[key]

Reviewer 2: This function was only useful in my repro to compare against; if it isn’t used in this snippet, it may just be confusing to have it there.

Author: True, thank you! we've simplified the snippets
